### PR TITLE
DATACOUCH-251 Wrap base where criteria with paranthesis

### PR DESCRIPTION
**Pull request by subalakr**
Thank you for proposing a pull request. This template will guide you through the essential steps necessary for a pull request.
Make sure that:

- [x] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [x] There is a ticket in the bug tracker for the project in our [JIRA](https://jira.spring.io/browse/DATACOUCH).
- [x] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).
- [x] You provide your full name and an email address registered with your GitHub account. If you’re a first-time submitter, make sure you have completed the [Contributor’s License Agreement form](https://support.springsource.com/spring_committer_signup).

Base criteria for the query should be wrapped before adding the entity filter
criteria so the priority of filtering is clear

### CODE REVIEW

File: src/main/java/org/springframework/data/couchbase/repository/query/support/N1qlUtils.java [hunk 1]
```diff
@@ -114,7 +115,7 @@ public static Expression createWhereFilterForEntity(Expression baseWhereCriteria
     if (baseWhereCriteria == null) {
       baseWhereCriteria = typeSelector;
     } else {
-      baseWhereCriteria = baseWhereCriteria.and(typeSelector);
+      baseWhereCriteria = x("(" + baseWhereCriteria.toString() + ")").and(typeSelector);
```
Comment (simonbasle) [inline src/main/java/org/springframework/data/couchbase/repository/query/support/N1qlUtils.java]: no better way of doing this, maybe `Expression.prefix/suffix/infix` could be made public in a future SDK release?
you could add a test that uses a `baseWhereCriteria` to `N1qlUtilsTest`:

```java
  @Test
  public void testCreateWhereFilterForEntityWithBaseCriteria() throws Exception {
    String expected = "(field >= 30 OR field2 = \"foo\") AND `_class` = \"java.lang.String\"";
    CouchbaseConverter converter = mock(CouchbaseConverter.class);
    when(converter.getTypeKey()).thenReturn("_class");
    EntityMetadata metadata = mock(EntityMetadata.class);
    when(metadata.getJavaType()).thenReturn(String.class);

    String real = N1qlUtils.createWhereFilterForEntity(
            x("field").gte(30).or(x("field2").eq(s("foo"))),
            converter, metadata).toString();

    assertEquals(expected, real);
  }
```

File: src/test/java/org/springframework/data/couchbase/repository/query/PartTreeN1qBasedQueryTest.java [hunk 2]
```diff
@@ -112,17 +114,50 @@ public void testGetCountExcludesDynamicSortClause() throws Exception {
...
+				+ " AND `_class` = \"org.springframework.data.couchbase.core.Beer\"", statement.toString());
+	}
+
 	public static interface TestRepository extends CrudRepository<Beer, String> {
 
 		Page<Beer> findByNameOrderByName(String name, Pageable pageRequest);
 
 		Page<Beer> findByName(String name, Pageable pageRequest);
 
+		List<Beer> findByIdOrName(String id, String name);
```
Comment (simonbasle) [inline src/test/java/org/springframework/data/couchbase/repository/query/PartTreeN1qBasedQueryTest.java]: I would probably also test in an integration test that that sort of query does indeed return the expected results from db when executed from end to end...

**Comment by simonbasle** [comment]
@subalakr I think you can omit the pull-request template now that you've signed the CLA, since you're the official main committer to the project now 😄

File: src/main/java/org/springframework/data/couchbase/repository/query/support/N1qlUtils.java (hunk 1 above)
Comment (subalakr) [inline src/main/java/org/springframework/data/couchbase/repository/query/support/N1qlUtils.java]: yes no other way for now, will expose an api to group expressions

**Comment by subalakr** [comment]
merged

**Pull request closed without merging**